airflow/jobs/scheduler_job.py

```diff
@@ -1706,62 +1342,326 @@ def _run_scheduler_loop(self) -> None:
+        # TODO[HA]: Rename update_state -> schedule_dag_run, ?? something else?
+        schedulable_tis = dag_run.update_state(session=session, execute_callbacks=False)
+        # TODO[HA]: Don't return, update these from in update_state?
+
+        # Get list of TIs that do not need to executed, these are
+        # tasks using DummyOperator and without on_execute_callback / on_success_callback
+        dummy_tis = [
+            ti for ti in schedulable_tis
+            if
+            (
+                ti.task.task_type == "DummyOperator"
+                and not ti.task.on_execute_callback
+                and not ti.task.on_success_callback
+            )
+        ]
+
+        # This will do one query per dag run. We "could" build up a complex
+        # query to update all the TIs across all the execution dates and dag
+        # IDs in a single query, but it turns out that can be _very very slow_
+        # see #11147/commit ee90807ac for more details
+        count = session.query(TI).filter(
+            TI.dag_id == dag_run.dag_id,
+            TI.execution_date == dag_run.execution_date,
+            TI.task_id.in_(ti.task_id for ti in schedulable_tis if ti not in dummy_tis)
+        ).update({TI.state: State.SCHEDULED}, synchronize_session=False)
+
+        # Tasks using DummyOperator should not be executed, mark them as success
+        if dummy_tis:
+            session.query(TI).filter(
+                TI.dag_id == dag_run.dag_id,
+                TI.execution_date == dag_run.execution_date,
+                TI.task_id.in_(ti.task_id for ti in dummy_tis)
+            ).update({
+                TI.state: State.SUCCESS,
+                TI.start_date: timezone.utcnow(),
+                TI.end_date: timezone.utcnow(),
+                TI.duration: 0
+            }, synchronize_session=False)
+
+        return count
+
+    @provide_session
+    def _verify_integrity_if_dag_changed(self, dag_run: DagRun, session=None):
+        """Only run DagRun.verify integrity if Serialized DAG has changed since it is slow"""
+        latest_version = SerializedDagModel.get_latest_version_hash(dag_run.dag_id, session=session)
+        if dag_run.dag_hash == latest_version:
+            self.log.debug("DAG %s not changed structure, skipping dagrun.verify_integrity", dag_run.dag_id)
+            return
+
+        dag_run.dag_hash = latest_version
+
+        # Refresh the DAG
+        dag_run.dag = self.dagbag.get_dag(dag_id=dag_run.dag_id, session=session)
+
+        # Verify integrity also takes care of session.flush
+        dag_run.verify_integrity(session=session)
+
+    def _send_dag_callbacks_to_processor(self, dag_run: DagRun):
+        if not self.processor_agent:
+            raise ValueError("Processor agent is not started.")
+
+        dag = dag_run.get_dag()
+        self._manage_slas(dag)
+        if dag_run.callback:
+            self.processor_agent.send_callback_to_execute(dag_run.callback)
+
+    def _manage_slas(self, dag: DAG):
```

Comment:
Is this name still correct? It seems to me that its content changed drastically while the name remained without drastic change.

Reply:
Can change it to `_send_sla_callbacks`, WDYT?

Reply:
Done in https://github.com/apache/airflow/pull/10956/commits/81129c252583e9dff54142eef1344749ea83c55b